module/mempool/herocache/backdata/heropool/linkedlist.go

```diff
@@ -2,13 +2,59 @@ package heropool
 
 // link represents a slice-based doubly linked-list node that
 // consists of a next and previous poolIndex.
+// if a link doesn't belong to any state it's next and prev may have any values,
+// but those are treated as invalid and should not be used.
 type link struct {
-	next poolIndex
-	prev poolIndex
+	next EIndex
+	prev EIndex
 }
 
 // state represents a doubly linked-list by its head and tail pool indices.
+// If state has 0 size, its tail's and head's prev and next are treated as invalid.
+// Moreover head's prev and tails next are always treated as invalid and may hold any values.
 type state struct {
-	head poolIndex
-	tail poolIndex
+	head EIndex
+	tail EIndex
+	size uint32
+}
+
+// adds entity to the tail of the list or creates first element
+func (s *state) appendEntity(p *Pool, entityIndex EIndex) {
+	if s.size == 0 {
+		s.head = entityIndex
+		s.tail = entityIndex
+		s.size = 1
+		return
+	}
+	p.connect(s.tail, entityIndex)
+	s.size++
+	s.tail = entityIndex
+}
+
+// removes an entity from the list
+func (s *state) removeEntity(p *Pool, entityIndex EIndex) {
+	if s.size == 0 {
+		panic("Removing an entity from the empty list")
+	}
+	if s.size == 1 {
+		s.size--
+		return
+	}
+	node := p.poolEntities[entityIndex].node
+
+	if entityIndex != s.head && entityIndex != s.tail {
+		// links next and prev elements for non-head and non-tail element
+		p.connect(node.prev, node.next)
+	}
+
+	if entityIndex == s.head {
+		// moves head forward
+		s.head = node.next
+	}
+
+	if entityIndex == s.tail {
+		// moves tail backwards
+		s.tail = node.prev
+	}
+	s.size--
```

Comment:
Thank you for detailed explanation. 

I would like to explain what I had in mind and propose several options as a follow up.

1. Put

Reply:
Thank you for your well-articulated input. After carefully considering the options presented, I recommend proceeding with Option (1). Moreover, during the implementation phase, I request that particular attention be paid to ensuring the minimization of additional independent heap objects introduced. This is imperative for the solution's optimal efficiency, correctness, and performance.

Reply:
sure, will do. But just as last point what I meant is something like 

type state struct {
	head EIndex
	tail EIndex
	size uint32
        nodes []link
}

and in parallel "node link" would have been removed from  poolEntity struct. So, like this there is only one extra allocation on the heap when the pool is created. So, each list would maintain its own array of nodes allocated on a construction of the pool.

Reply:
There are 3 things that I'd like to discuss.

1. As -1 is now an undefined value then shell we assign it to all nodes prev, and next on a pool initialization ?
In the current solution those values are 0 which is fine as by convention those are the invalid ones. In my first proposal there are no invalid values so 0s are ok. However now probably these should be set to -1.
2. Also, when freeing a node from any list might be good then to set it's prev and next to -1.
3. I noticed that for the "const" methods (p Pool) copy of the pool is used. I wonder if those copies have impact on cache efficiency. It looks to be minor as I read that Go would not make a deep copy of an array included in the pool, but still read that a pointer would be more efficient. Not sure if it's worth though.

Reply:
I would strongly recommend not proceeding with the proposal to relocate the "node link" from `poolEntity` to `state`. There are two fundamental reasons for my caution:

1. Within `HeroCache`, memory allocation for the entire `Pool` is carried out during initialization using an array structure. In the present design, this is achieved in a singular location within `Pool`, namely `poolEntities: make([]poolEntity, sizeLimit),`. The proposed alteration, however, necessitates extending memory initialization to two additional areas by initializing the `free` and `used` states, where the `nodes` attribute must also be initialized with an identical `sizeLimit`. Although, as you pointed out, the impact on heap memory is negligible, this change would inadvertently render the code somewhat less maintainable and marginally more susceptible to errors. Specifically, any discrepancy in the sizes of these three elements could potentially introduce a bug.

2. Additionally, the suggested change has the unintended consequence of making the system more vulnerable to an edge case where a pool entity is erroneously included in both the `free` and `used` lists. In the current design, this edge case is restricted to the heads and tails, but by transferring link management to the `state`, there is a heightened risk that the `free` and `used` states may overlap on the same pool entity.

Regarding your inquiries:

1. Please refer to my response in comment https://github.com/onflow/flow-go/pull/4470#discussion_r1237672741. In alignment with what is stated there, it is indeed necessary for the `prev` and `next` of all nodes to be set to `UndefinedEIndex` upon pool initialization.

2. Absolutely, when releasing a node from any list, it remains prudent to assign its `prev` and `next` values to `UndefinedEIndex`.

3. I appreciate you drawing attention to this issue. It has been acknowledged that this change was made inadvertently, and measures are being taken to address it through a separate pull request.

Reply:
For first 2 points:

Interesting point 1 and 2.  I see your point, but if the idea is to move to multiple lists and eviction strategy being an independent type/interface I think it would be nice to decouple lists from pool. If the size is set only once I think the risk for bugs is low. But surely more error prone rather than current solution. So, thank you for taking time to explain your point of view. Now, I understand better general idea.

Inquiries:

1. will do, thanks
2. same here, will do.
3. understood, happy that could help.

Reply:
I agree with your point-1 regarding moving the loop prior to the call on `l.initFreeEntities()`. This adjustment serves as a defensive mechanism to prevent illegal states and guarantees that `initFreeEntities` will fully process undefined links. While the current code appears to be functioning correctly, incorporating this loop would result in a more consistent implementation aligned with the invariants of the `HeroCache`. Furthermore, it is an inexpensive operation that only executes during initialization, making it a valuable addition.

Reply:
understood, done, resolving the thread.